### PR TITLE
fix(snippet): verify before get session info

### PR DESCRIPTION
**Pull request by axdank**
Error with default order:

### CODE REVIEW

**Comment by mfussenegger** [comment]
Could you please also add a test case or an explanation of how you produce the error?

File: runtime/lua/vim/snippet.lua [hunk 1]
```diff
@@ -346,16 +350,17 @@ local function setup_autocmds(bufnr)
...
+        return true
+      end
+
       -- Check that the snippet hasn't been deleted.
       local snippet_range = get_extmark_range(M._session.bufnr, M._session.extmark_id)
       if
         (snippet_range[1] == snippet_range[3] and snippet_range[2] == snippet_range[4])
         or snippet_range[3] + 1 > vim.fn.line('$')
       then
         M.exit()
```
Comment (MariaSolOs) [inline runtime/lua/vim/snippet.lua:363]: `M.exit` will clear the autocommands, so we don't need to return `true` here.

File: runtime/lua/vim/snippet.lua [hunk 2]
```diff
@@ -307,6 +307,10 @@ local function setup_autocmds(bufnr)
         return
       end
 
+      if not M.active() then
+        return true
+      end
+
```
Comment (MariaSolOs) [inline runtime/lua/vim/snippet.lua:313]: Might as well move this block before the `mode` check.

**Comment by MariaSolOs** [comment]
> Check now that I don't have a keymap associated with Esc in `cmp`, therefore I never call `vim.snippet.exit()`. I think that is the main problem generate this situation.

There's still a basic heuristic for cancelling the session when leaving the snippet region (but clearly we're missing some cases). I would prefer making us that heuristic more robust instead of adding a bunch of if-checks everywhere.

**Pull request closed without merging**